### PR TITLE
Push down filters to ParquetFileFormat

### PR DESCRIPTION
**Pull request by LuciferYang**
## What changes were proposed in this pull request?
As ParquetFileFormat, pushdown filters for skip parquet rowgroups when task degenerate to full scan.

#536 task 3

It use by DefaultOapRecordReader & VectorizedOapRecorder
 
## How was this patch tested?
FilterHelperSuite

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelper.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFilters
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+
+object FilterHelper {
+
+  def tryToPushFilters(sparkSession: SparkSession,
+                       requiredSchema: StructType,
+                       filters: Seq[Filter]): Option[FilterPredicate] =
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelper.scala]: indent

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelperSuite.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      .add(StructField("a", IntegerType))
+      .add(StructField("b", StringType))
+    val filters = Seq(GreaterThan("a", 1), EqualTo("b", "2"))
+    val pushed = FilterHelper.tryToPushFilters(filterPushDown = false, requiredSchema, filters)
+    assert(pushed.isEmpty)
+    val config = new Configuration()
+    FilterHelper.setFilterIfExist(config, pushed)
+    assert(config.get(ParquetInputFormat.FILTER_PREDICATE) == null)
+    assert(config.get(ParquetInputFormat.FILTER_PREDICATE + ".human.readable") == null)
+
```
Comment (linhong-intel) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelperSuite.scala]: remove empty line

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelperSuite.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    val expected = s"""and(gt(a, 1), eq(b, Binary{"2"}))"""
+    val pushed = FilterHelper.tryToPushFilters(filterPushDown = true, requiredSchema, filters)
+    assert(pushed.isDefined)
+    assert(pushed.get.toString.equals(expected))
+    val config = new Configuration()
+    FilterHelper.setFilterIfExist(config, pushed)
+    val humanReadable = config.get(ParquetInputFormat.FILTER_PREDICATE + ".human.readable")
+    assert(humanReadable.nonEmpty)
+    assert(humanReadable.equals(expected))
+
```
Comment (linhong-intel) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelperSuite.scala]: remove empty line

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelper.scala [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    tryToPushFilters(sparkSession.sessionState.conf.parquetFilterPushDown, requiredSchema, filters)
+
+  def tryToPushFilters(filterPushDown: Boolean,
+                       requiredSchema: StructType,
+                       filters: Seq[Filter]): Option[FilterPredicate] = {
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelper.scala]: indent

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 5]
```diff
@@ -303,9 +304,11 @@ private[sql] class OapFileFormat extends FileFormat
               Iterator.empty
             case _ =>
               OapIndexInfo.partitionOapIndex.put(file.filePath, false)
+              FilterHelper.setFilterIfExist(conf, pushed)
               val reader = new OapDataReader(
                 new Path(new URI(file.filePath)), m, filterScanners, requiredIds)
               val iter = reader.initialize(conf, options)
+              Option(TaskContext.get()).foreach(_.addTaskCompletionListener(_ => iter.close()))
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: @linhong-intel @TigerSong why remove this line from #580 ?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelperSuite.scala (hunk 2 above)
Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelperSuite.scala]: ok~

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelperSuite.scala (hunk 3 above)
Comment (LuciferYang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelperSuite.scala]: ok~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelper.scala (hunk 4 above)
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/utils/FilterHelper.scala]: ok~ format it

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala (hunk 5 above)
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: I don't think it should be removed...

Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: Thanks @LuciferYang  This is my mistake

**Comment by linhong-intel** [comment]
@LuciferYang conflict... :)

**Comment by LuciferYang** [comment]
@linhong-intel alrealy fix it ~

**Pull request merged**